tests/test_fibers.cpp

```
@@ -0,0 +1,168 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <alloca.h>
+#include <pthread.h>
+
+#include <rocket/rocket_engine.h>
+#include <rocket/rocket_executor.h>
+#include <rocket/rocket_fiber.h>
+
+#include <gtest/gtest.h>
+
+static const size_t queue_depth = 10;
+
+typedef struct {
+  size_t count;
+  size_t max;
+  size_t workers;
+} count_context_t;
+
+// Worker function used by the round-robin scheduling test case.
+static void* count_yield_worker(void* context) {
+  count_context_t* count_context = (count_context_t*)context;
+  while (count_context->count < count_context->max) {
+    // Increment the counter and snapshot the previous value.
+    const size_t last = ++count_context->count;
+
+    // Yield to the next fiber to run.
+    rocket_fiber_yield();
+
+    /* Other workers should have incremented the counter since we yielded, but
+     * never beyond the maximum.
+     */
+    const size_t expected_count =
+        std::min(last + count_context->workers - 1, count_context->max);
+    EXPECT_EQ(expected_count, count_context->count);
+  }
+  return nullptr;
+}
+
+/* Test case to verify that work items are scheduled round-robin when they
+ * manually yield and are not blocked on IO.
+ */
+static void test_round_robin_scheduling(void) {
+  rocket_engine_t* engine = rocket_engine_create(queue_depth);
+  EXPECT_NE(engine, nullptr);
```

IIRC, `EXPECT_*` will continue the test even after a failed expectation where as `ASSERT_*` fails fast and stops the test right away. In case the engine or the executor is unfortunately `nullptr`, I feel it makes more sense to fail fast here.